### PR TITLE
Add Tree Query Language and Engine

### PR DESCRIPTION
**Pull request by AntonyBlakey**
This pull request adds the CST Tree Query Language.

It is a long way from being optimised, and in particular it should not need to allocate memory whilst backtracking, which it currently does.

It is built with understandability as the top priority.

This also includes some other changes that are not quite drive-by e.g. I rationalised the runtime crate structure and file naming to make more sense given I was adding a structural sub-module.

### CODE REVIEW

File: .changeset/curvy-donkeys-shout.md
Comment (OmarTawfik) [inline .changeset/curvy-donkeys-shout.md:5]: I wonder if this is a `minor` bump, given it is a new feature.

File: crates/codegen/parser/runtime/src/templates/kinds.rs.jinja2
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/templates/kinds.rs.jinja2:16]: `phf` crate looks like a great addition! but I wonder if it is required for the query engine.
If not, what do you think about adding it separately, possibly once we have a perf suite, so that we can measure/report a realistic delta (for this use case, and possibly others)?

File: crates/codegen/parser/runtime/src/cursor.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/cursor.rs:67]: I wonder if the only usage for this is the one in `matches_query_node_id()` .. If so, what do you think of adding a `Cursor::node_ref()` similar to `Cursor::node()` that returns a `&Node` instead of cloning .. I imagine it is useful for the public API even .. as many user operations don't require cloning .. instead of exposing these implementation details (`Cursor::current` and `PathNode`) ..

File: crates/codegen/parser/runtime/src/kinds.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/kinds.rs:22]: nit: I suggest moving/rewording the older comment, as it now precedes the wrong node. Maybe moving it to the `TokenKind` enum type doc comments?

File: crates/codegen/parser/runtime/src/lib.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/lib.rs:12]: Thouhgts about keeping the new (4) modules under a parent modue `query/mod.rs`?

File: crates/solidity/outputs/npm/crate/src/generated/napi/napi_ast_selectors.rs
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/crate/src/generated/napi/napi_ast_selectors.rs:2]: looks like this was added during rebase? it was already deleted earlier.

File: Cargo.toml
Comment (OmarTawfik) [inline Cargo.toml:80]: nit: I suggest keeping the list alphabetically sorted.

File: package-lock.json
Comment (OmarTawfik) [inline package-lock.json:81]: I think this might have been added during rebase?

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_model.rs:83]: What do you think of using "text" instead of "String", which is more aligned with `Token::text`:

```suggestion
    Anonymous,
    Kind { kind: Kind },
    Text { text: String },
    Field { field: FieldName },
    FieldAndKind { field: FieldName, kind: Kind },
    FieldAndText { field: FieldName, text: String },
```

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_model.rs:13]: What do you think of using the actual semantics (similar to other variants) rather than the syntax kind? for example:

```suggestion
    ZeroOrMoreNodes,
```

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_model.rs:62]: instead of exposing `query_model::Kind` here, as we might need that type in the future for similar reasons, what do you think about exposing it publicly as `kinds::NodeKind` instead?
This is similar to the `kinds.NodeKind` TypeScript type we already export.

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_model.rs:21]: TOL, since `unparse()` doesn't store/reconstruct the input query, what do you think about using something like `impl Display`>`to_string()` instead?
Would also make it a bit more idiomatic/effecient, as it uses a single writer instead of creating/dropping temporary strings.

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_model.rs:77]: nit suggestion: `NodeId` was a bit confusing to me, considering we have a node kind, a node text/string.
Thoughts about using something like `NodeSelector`?

File: crates/codegen/parser/runtime/src/query_parser_tests.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_parser_tests.rs:9]: I wonder about the long term plan for such tests .. will they use something like `cst_snapshots` to write `generated/*` files?

If not, in order to assert the actual model parsed, what do you think about using something like [inline snapshots](https://docs.rs/insta/latest/insta/#inline-snapshots) to generate/update them in the same file?

File: crates/codegen/parser/runtime/src/query_parser_tests.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_parser_tests.rs:10]: I suggest adding a separate test for each query/variant .. to test it e2e.

File: crates/codegen/parser/runtime/src/query_engine.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_engine.rs:57]: TOL, I'm a bit confused by using `QecRef` (`QueryCombinator`?) vs `BindingCombinator` vs `QueryEngineCombinator` vs `QueryIterator` .. but maybe I need to spend some time using these types to have a better understanding ..

File: crates/codegen/parser/runtime/src/query_engine_tests.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_engine_tests.rs:87]: thoughts about using Solidity code snippets as input vs `cst_tree!()`?

File: crates/codegen/parser/runtime/src/query_engine_tests.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query_engine_tests.rs:159]: I suggest adding tests for the rest of the constructs, for example, filtering by field names..

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
Left a few questions/suggestions.
Otherwise, this looks awesome! 🎉

File: crates/codegen/parser/runtime/src/cursor.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/cursor.rs:67]: I think that an accessors such as `node()` should return a reference, and not clone a value. That is contrary to my expectation about accessors. However I want to talk about that with the three of us, so for now I will use the accessor and pretend it is a reference.

File: crates/codegen/parser/runtime/src/templates/kinds.rs.jinja2
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/templates/kinds.rs.jinja2:16]: I originally put this in because I was doing the string -> enum conversion in an inner loop, but then I took that out, and so this wasn't needed. But I decided to leave it in because that operation is likely to be used by clients, and they have no way of doing this optimisation themselves because it needs to be applied here. Plus I think that phf should be the default way this is implemented.

The cost is small, and even though for *our* current code it is an unneccesary optimisation, unjustified by profiling, we cannot say the same for client code, and only we can make this choice.

File: crates/codegen/parser/runtime/src/kinds.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/kinds.rs:22]: Done

File: crates/codegen/parser/runtime/src/query_engine.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_engine.rs:57]: QueryEngineCombinatorRef

Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_engine.rs:57]: Open to suggestions. This is about as simple as I could get, and it took months to get here ... but always happy to improve naming. Maybe I'm being too concise, and the names should be more explicit?

File: crates/codegen/parser/runtime/src/query_engine_tests.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_engine_tests.rs:87]: Originally I thought the same thing, but in this crate we have no solidity :)

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_model.rs:13]: Well, Ellipsis is a semantic operator. I think you are meaning the implementation semantics, and in this case I think Ellipsis is better, but it's a minor preference over `Anything`

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_model.rs:21]: Good reasoning, done

File: crates/codegen/parser/runtime/src/query_parser_tests.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_parser_tests.rs:10]: I think a few edge case tests that show how the input is parsed and then unparsed to a canonical structure is all that is needed, given the engine takes query strings as test inputs and should be more thorough.

File: package-lock.json
Comment (AntonyBlakey) [reply package-lock.json:81]: Done

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_model.rs:62]: YAGNI? I don't think we should hoist this in anticipation of maybe wanting it. I will look at limiting the visibility

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_model.rs:77]: Yes, good suggestion.

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_model.rs:83]: Yes, good suggestion.

File: crates/codegen/parser/runtime/src/query_parser_tests.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_parser_tests.rs:9]: I don't see the benefit in using snapshots here, I think traditional tests are fine, especially, but not only because  we don't need many tests here because the engine tests also coincidentally test the parser.

File: crates/codegen/parser/runtime/src/lib.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/lib.rs:12]: I was in two minds about this. So I guess that's 1.5 vs 0.5, so will do.

File: Cargo.toml
Comment (AntonyBlakey) [reply Cargo.toml:80]: Done

File: crates/codegen/parser/runtime/src/query_engine.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_engine.rs:57]: I changed to `Combinator` and `CombinatorRef`, because the types are local to the module and the names can therefore be scoped.

File: crates/solidity/outputs/npm/crate/src/generated/napi/napi_ast_selectors.rs
Comment (AntonyBlakey) [reply crates/solidity/outputs/npm/crate/src/generated/napi/napi_ast_selectors.rs:2]: done

**Review by AntonyBlakey — COMMENTED** [review]
> I suggest adding tests for the rest of the constructs, for example, filtering by field names.. 

Done

File: crates/codegen/parser/runtime/src/query_engine_tests.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_engine_tests.rs]: Done

File: crates/codegen/parser/runtime/src/query_parser_tests.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query_parser_tests.rs]: Done

File: crates/codegen/parser/runtime/src/kinds.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/kinds.rs:21]: nit/spelling:

```suggestion
    // Used for testing in this crate. This is generated in the client code.
```

File: crates/codegen/parser/runtime/src/query_engine.rs
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/query_engine.rs]: Sounds great! thanks1

File: crates/codegen/parser/runtime/src/query_engine_tests.rs
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/query_engine_tests.rs]: Oh right 😅

Sounds great. thanks for explaining.

File: crates/codegen/parser/runtime/src/query_model.rs
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/query_model.rs]: Not sure if YAGNI applies here?
I mean, if we are doing the work regardless of defining and maintaining that type, I suggest exposing it through the `cst` module instead (for organizational/above reasons).
If we can limit its visibility/there is no need to expose it, that also sounds good to me.

File: crates/codegen/parser/runtime/src/kinds.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/kinds.rs:23]: nit: what do you think about using `Token1`, `Token2`, `Token3` instead of `X`, `Y`, and `Z`?
And `Rule1`, `Rule2`, `Rule3` instead of `A`, `B`, `C`?

I think it would make it a bit clearer/easier to me to understand the query tests, if the variant we explicit here. It is worth a few extra characters.

File: crates/codegen/parser/runtime/src/query_parser_tests.rs
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/query_parser_tests.rs]: at least a simple basic test for each syntax feature/variant we have, to protect against accidental regressions/changes in the future.

File: crates/codegen/parser/runtime/src/templates/kinds.rs.jinja2
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/templates/kinds.rs.jinja2:16]: > But I decided to leave it in because that operation is likely to be used by clients, and they have no way of doing this optimisation themselves because it needs to be applied here. Plus I think that phf should be the default way this is implemented.

fully agree .. but if it is unrelated to/required by the query engine, I suggest doing it separately.

And I agree that it is going to be an improvement. But if we have a perf suite, it is an excellent use case/test case to use for it. Not to justify it, but to prove how much we care and affect performance by the work we are currently doing.

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query/engine.rs:17]: not sure I understand the difference between `Cursor::consume()` and `Cursor::complete()` from the names. The first only completes the current node, right? it won't actually consume the cursor, and it doesn't have to be completed afterwards.

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query/engine.rs:28]: nit spelling:

```suggestion
    fn matches_node_selector(&self, node_selector: &NodeSelector) -> bool {
```

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query/engine.rs:34]: nit: leftover from the rename:

.. also others ..

```suggestion
                NodeSelector::FieldName { field_name } => Some(*field_name) == self.node_name(),
```

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/query/engine.rs:242]: not sure I'm following the logic here. For example, why are we returning a cloned/consumed cursor?
Happy to wait for future docs if that answers such questions.

**Review by OmarTawfik — APPROVED** [review]

**Review by OmarTawfik — APPROVED** [review]
PR looks great. Accepted to unblock you, but there are a couple of remaining concerns I would suggest addressing before merging (phf, and the cursor/query public API). The others are just spelling/nits.

File: crates/codegen/parser/runtime/src/kinds.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/kinds.rs:23]: Done

File: crates/codegen/parser/runtime/src/kinds.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/kinds.rs:21]: Done

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query/engine.rs:34]: Done

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query/engine.rs:28]: Done

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query/engine.rs:29]: For completely private methods like these, I think they can go on whatever type is most regular for the problem being solved. I was in two minds about this, but came down on the cursor being the right place for the sense of the algorithm.

Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query/engine.rs:29]: As for the `query` method itself, queries are layered on top of cursors - they add to the functionality but don't belong on the 'canonical' definition of the type. I didn't feel the need to make a `Queryable` trait, but maybe that would make sense, and then it could be on `Node` as well. WDYT?

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query/engine.rs:17]: Renamed to make it more obvious. The tricky bit is mixing the return value and the action. This is specific to the query execution semantics. I put it here for DRY.

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/query/engine.rs:242]: This is for symmetry wth the snippet above, but good catch!

File: crates/codegen/parser/runtime/src/query/engine.rs
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/query/engine.rs:29]: Since that Rust allows defining extension methods on either types (modules), I think it is more about the UX when importing and using a `slang::cursor::Cursor` .. if they want to call a `Cursor::query()` method:

- if there is a single possible query that can be returned, and it is not one of many implementations/variations of what a `Query` is .. then I think it should live on the `Cursor` type, because it is a unique API bound only to that type, even if the cursor can be used without the query .. they don't have to import the query module, or know about its types ..
- On the other hand, if we make a `Queryable` trait, then the trait would naturaly live here in the query module, and its implementations would live in the cursor module (and also the `cst` module as you suggested).

Makes sense that a node would be queryable .. if the extra implementation is not too much work, I would say let's do it.

**Review by OmarTawfik — APPROVED** [review]

**Pull request merged**
